cpp/src/arrow/c/dlpack.cc

```diff
@@ -0,0 +1,151 @@
+    case Type::UINT32:
+    case Type::UINT64:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      *out = dtype;
+      return Status::OK();
+    case Type::HALF_FLOAT:
+    case Type::FLOAT:
+    case Type::DOUBLE:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      *out = dtype;
+      return Status::OK();
+    case Type::BOOL:
+      // DLPack supports byte-packed boolean values
+      return Status::TypeError("Bit-packed boolean data type not supported by DLPack.");
+    default:
+      return Status::TypeError(
+          "Can only use __dlpack__ on primitive arrays without NullType and Decimal "
+          "types.");
+  }
+}
+
+struct DLMTensorCtx {
+  std::shared_ptr<ArrayData> ref;
+  std::vector<int64_t> shape;
+  DLManagedTensor tensor;
+};
+
+static void deleter(DLManagedTensor* arg) {
+  delete static_cast<DLMTensorCtx*>(arg->manager_ctx);
+}
+
+Status ExportArray(const std::shared_ptr<Array>& arr, DLManagedTensor** out) {
+  if (arr->null_count() > 0) {
+    return Status::TypeError("Can only use __dlpack__ on arrays with no nulls.");
+  }
+
+  // Define the DLDataType struct
+  // Supported data types: int, uint, float
+  DLDataType arr_type;
+  RETURN_NOT_OK(getDLDataType(arr->type(), &arr_type));
+
+  // Create DLMTensorCtx struct with the reference to
+  // the data of the array
+  std::shared_ptr<ArrayData> array_ref = arr->data();
+  DLMTensorCtx* DLMTensor = new DLMTensorCtx;
+  DLMTensor->ref = array_ref;
+
+  // Define DLManagedTensor struct defined by
+  // DLPack (dlpack_structure.h)
+  DLManagedTensor* dlm_tensor = &DLMTensor->tensor;
+  dlm_tensor->manager_ctx = DLMTensor;
+  dlm_tensor->deleter = &deleter;
+
+  // Define the data pointer to the DLTensor
+  // If array is of length 0, data pointer should be NULL
+  if (arr->length() == 0) {
+    dlm_tensor->dl_tensor.data = NULL;
+  } else if (arr->offset() > 0) {
+    const auto byte_width = arr->type()->byte_width();
+    const auto start = arr->offset() * byte_width;
+    ARROW_ASSIGN_OR_RAISE(auto sliced_buffer,
+                          SliceBufferSafe(array_ref->buffers[1], start));
+    dlm_tensor->dl_tensor.data =
+        const_cast<void*>(reinterpret_cast<const void*>(sliced_buffer->address()));
+  } else {
+    dlm_tensor->dl_tensor.data = const_cast<void*>(
+        reinterpret_cast<const void*>(array_ref->buffers[1]->address()));
+  }
+
+  // Define DLDevice struct
+  DLDevice ctx;
+  if (array_ref->buffers[1]->device_type() == DeviceAllocationType::kCPU) {
+    ctx.device_id = 0;
+    ctx.device_type = DLDeviceType::kDLCPU;
+  } else {
+    return Status::NotImplemented(
+        "DLPack support is implemented only for buffers on CPU device.");
+  }
+  dlm_tensor->dl_tensor.device = ctx;
+
+  dlm_tensor->dl_tensor.ndim = 1;
+  dlm_tensor->dl_tensor.dtype = arr_type;
+  std::vector<int64_t>* shape_arr = &DLMTensor->shape;
+  shape_arr->resize(1);
+  (*shape_arr)[0] = arr->length();
+  dlm_tensor->dl_tensor.shape = shape_arr->data();
```

Comment:
This is not necessary, we could simply point to the existing length and remove `DLMTensorCtx::shape`:
```suggestion
  dlm_tensor->dl_tensor.shape = const_cast<int64_t*>(&array_ref->length);
```